File: Floyd-Warshall/9205/9205.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <cmath>
using namespace std;
#define MAX 100
#define INF 999999999
typedef pair<int, int> pii;
int t, n, dp[MAX + 2][MAX + 2];
void init() {
    cin.tie(nullptr);
    cout.tie(nullptr);
    ios_base::sync_with_stdio(0);
}

int calDist(pii u, pii v) {
    int x1 = u.first, y1 = u.second;
    int x2 = v.first, y2 = v.second;
    int dist = abs(x1 - x2) + abs(y1 - y2);
    return dist;
}

int floyd(vector<pii> coor) {
    fill(&dp[0][0], &dp[MAX][MAX + 1], INF);
    for (int u = 0; u < n + 2; u++) {
        for (int v = 0; v < n + 2; v++) {
            if (u == v) continue;
            int dist = calDist(coor[u], coor[v]);
            if (dist <= 1000) {
                dp[u][v] = 1;
            }
        }
    }
    for (int k = 0; k < n + 2; k++) {
        for (int u = 0; u < n + 2; u++) {
            for (int v = 0; v < n + 2; v++) {
                dp[u][v] = min(dp[u][v], dp[u][k] + dp[k][v]);
            }
        }
    }
    return dp[0][n + 1];
}

int main() {
    init();
    cin >> t;
    while (t--) {
        cin >> n;
        int x, y;
        cin >> x >> y; //집
        vector<pii> coor;
        coor.push_back(pii(x, y));
        for (int iter = 0; iter < n; iter++) {
            cin >> x >> y; //편의점
            coor.push_back(pii(x, y));
        }
        cin >> x >> y; //도착지
        coor.push_back(pii(x, y));
        int ans = floyd(coor);
        if (ans == INF) cout << "sad\n";
        else cout << "happy\n";
    }
    return 0;
}

File: Floyd-Warshall/9205/9205.kt
import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter
import kotlin.math.abs
import kotlin.math.min

private const val INF = 999999999
fun main(args: Array<String>) = with(BufferedReader(InputStreamReader(System.`in`))) {
    val bw = BufferedWriter(OutputStreamWriter(System.out))
    val points = arrayListOf<Pair<Int, Int>>()

    fun calDist(u: Pair<Int, Int>, v: Pair<Int, Int>): Int {
        return abs(u.first - v.first) + abs(u.second - v.second)
    }

    fun floyd(n: Int) {
        val dp = Array(n + 2) { IntArray(n + 2) { INF } }

        repeat(n + 2) { u ->
            repeat(n + 2) { v ->
                if (u == v) return@repeat
                val dist = calDist(points[u], points[v])
                if (dist <= 1000) dp[u][v] = 1
            }
        }

        repeat(n + 2) { k ->
            repeat(n + 2) { u ->
                repeat(n + 2) { v ->
                    dp[u][v] = min(dp[u][v], dp[u][k] + dp[k][v])
                }
            }
        }

        bw.write(if (dp[0][n + 1] == INF) "sad\n" else "happy\n")
    }

    val t = readLine().toInt()
    repeat(t) {
        points.clear()

        val n = readLine().toInt()
        var point = readLine().split(" ").map { it.toInt() }
        points.add(Pair(point[0], point[1])) // 집

        repeat(n) { //편의점
            point = readLine().split(" ").map { it.toInt() }
            points.add(Pair(point[0], point[1]))
        }

        point = readLine().split(" ").map { it.toInt() }
        points.add(Pair(point[0], point[1])) // 도착지

        floyd(n)
    }

    bw.flush()
    bw.close()
}
